README.md
# game-of-life

[![Build Status](https://travis-ci.com/jackdbd/game-of-life.svg?branch=master)](https://travis-ci.org/jackdbd/game-of-life)

Conway's Game of Life implemented in Clojure and animated with [Quil](https://github.com/quil/quil).

![Demo](https://raw.githubusercontent.com/jackdbd/game-of-life/master/demo.gif "Demo GIF of the Game of Life")

## Usage

This project uses [Leiningen](https://leiningen.org/).

Play with:

```sh
lein run
```

## Build

Build the [uberjar](https://imagej.net/Uber-JAR):

```sh
lein uberjar
```

Then run the executable:

```sh
java -jar target/uberjar/game-of-life-standalone.jar
```

## Credits

- [Conway's Game of Life by Christophe Grand](http://clj-me.cgrand.net/2011/08/19/conways-game-of-life/)


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2020-05-25
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2020-05-25
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/game-of-life/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/game-of-life/compare/0.1.0...0.1.1


test/game_of_life/conway_test.clj
(ns game-of-life.conway-test
  (:require [clojure.test :refer [deftest is testing]]
            [game-of-life.conway :refer [neighbours step will-live?]]))

(deftest will-live?-alive-cell-test
  (testing "An alive cell will die if it has 1 neighbour"
    (is (= false (will-live? 1 [0 0]))))
  (testing "An alive cell will live if it has 2 neighbours"
    (is (= true (will-live? 2 [0 0]))))
  (testing "An alive cell will live if it has 3 neighbours"
    (is (= true (will-live? 2 [0 0]))))
  (testing "An alive cell will die if it has 4 neighbours"
    (is (= false (will-live? 4 [0 0])))))

(deftest will-live?-dead-cell-test
  (testing "A dead cell will stay dead if it has 1 neighbour"
    (is (= false (will-live? 1 nil))))
  (testing "A dead cell will stay dead if it has 2 neighbours"
    (is (= false (will-live? 2 nil))))
  (testing "A dead cell will come to life if it has 3 neighbours"
    (is (= true (will-live? 3 nil))))
  (testing "An alive cell will stay dead if it has 4 neighbours"
    (is (= false (will-live? 4 nil)))))

(deftest neighbours-test
  (testing "Any cell has 8 neighbours."
    (is (= 8 (count (neighbours [0 0]))))
    (is (= 8 (count (neighbours [1 3]))))
    (is (= 8 (count (neighbours [123 456]))))))

; https://en.wikipedia.org/wiki/Conway%27s_Game_of_Life#Examples_of_patterns
(deftest blinker-test
  (let [blinker #{[2 1] [2 2] [2 3]}]
    (testing "A blinker alternates between 2 states."
      (is (= blinker (step (step blinker)))))))

(deftest beehive-test
  (let [beehive #{[2 1] [1 2] [1 3] [2 4] [3 2] [3 3]}]
    (testing "A beehive never changes state."
      (is (= beehive (step beehive)))
      (is (= beehive (step (step beehive))))
      (is (= beehive (step (step (step beehive))))))))

(defn sorted-cells-by-xy-sum
  [cells]
  (sort (map (fn [cell] [(apply + cell) cell]) cells)))

; https://www.conwaylife.com/wiki/Glider
(deftest glider-test
  (let [glider #{[1 2] [2 3] [3 1] [3 2] [3 3]}
        n (count glider)
        generations (take 5 (iterate step glider))]
    (testing "A glider keeps the same number of alive cells across generations."
      (is (= 5 (count generations)))
      (is (= n (count (nth generations 2))))
      (is (= n (count (nth generations 3))))
      (is (= n (count (nth generations 4))))
      (is (= n (count (last generations)))))
    (testing "A glider travels 1 cell diagonally downward from generation 0 to 4."
      (let [[xy-sum-0 cell-0] (last (sorted-cells-by-xy-sum glider))
            glider-4 (last generations)
            [xy-sum-4 cell-4] (last (sorted-cells-by-xy-sum glider-4))]
        (is (= 6 xy-sum-0))
        (is (= [3 3] cell-0))
        (is (= 8 xy-sum-4))
        (is (= [4 4] cell-4))))))


project.clj
(defproject game-of-life "0.1.0-SNAPSHOT"
  :description "Conway's Game of Life with Clojure and Quil."
  :url "https://github.com/jackdbd/game-of-life"
  :license {:name "EPL-2.0 OR GPL-2.0-or-later WITH Classpath-exception-2.0"
            :url "https://www.eclipse.org/legal/epl-2.0/"}
  :dependencies [[org.clojure/clojure "1.10.1"]
                 [quil "3.1.0"]]
  :main game-of-life.core
  :target-path "target/%s"
  :uberjar-name "game-of-life-standalone.jar"
  :aliases {"format" ["with-profile" "dev" "cljfmt" "fix"]
            "lint" ["with-profile" "dev" "eastwood"]
            "static-analysis" ["with-profile" "dev" "kibit"]
            "fix" ^{:doc "Format code, lint code, perform static analysis."}
            ["do" ["format"] ["lint"] ["static-analysis"]]}
  :profiles {:uberjar {:aot :all}
             :dev {:dependencies [[io.aviso/pretty "0.1.37"]
                                  [pjstadig/humane-test-output "0.10.0"]]
                   :injections [(require 'pjstadig.humane-test-output)
                                (pjstadig.humane-test-output/activate!)]
                   :middleware [io.aviso.lein-pretty/inject]
                   :plugins [[com.jakemccrary/lein-test-refresh "0.24.1"]
                             [io.aviso/pretty "0.1.37"]
                             [jonase/eastwood "0.3.10"]
                             [lein-cljfmt "0.6.7" :exclusions [org.clojure/clojure]]
                             [lein-kibit "0.1.8"]]}})


src/game_of_life/core.clj
(ns game-of-life.core
  "Conway's Game of Life implemented in Clojure and animated with Quil."
  (:gen-class)
  (:require
   [game-of-life.ui :as ui]
   [quil.core :as q]
   [quil.middleware :as m]))

; https://github.com/quil/quil/wiki/Dynamic-Workflow-(for-REPL)
; http://quil.info/api/environment#defsketch
;; (q/defsketch game-of-life
;;   :draw ui/draw-fn
;;   :features [:keep-on-top]
;;   :middleware [m/pause-on-error m/fun-mode]
;;   :mouse-clicked ui/mouse-clicked
;;   :mouse-moved ui/mouse-moved
;;   :mouse-wheel ui/mouse-wheel
;;   :settings #(q/smooth 2)
;;   :setup ui/setup-fn
;;   :size [800 600]
;;   :title "Conway's Game of Life"
;;   :update ui/update-fn)

; https://github.com/quil/quil/wiki/Runnable-jar
(defn -main []
  (q/sketch
   :draw ui/draw-fn
   :features [:resizable :exit-on-close]
   ;; :features [:present :exit-on-close]
   :key-pressed ui/key-pressed
   :middleware [m/fun-mode]
   :mouse-clicked ui/mouse-clicked
   :mouse-moved ui/mouse-moved
   :mouse-wheel ui/mouse-wheel
   :setup ui/setup-fn
   :size [800 600]
   ;; :size :fullscreen
   :title "Conway's Game of Life"
   :update ui/update-fn))


src/game_of_life/conway.clj
(ns game-of-life.conway
  "Rules for the Game of Life.")

(defn neighbours
  "Moore neighborhood of a cellular automata."
  [[x y]]
  (for [dx [-1 0 1]
        dy (if (zero? dx) ; because a cell is not a neighour of itself
             [-1 1]
             [-1 0 1])]
    [(+ x dx) (+ y dy)]))

(defn will-live?
  "Test whether the cell will live or not.
  The Game of Life rules state that a cell will live if:
  - it has 3 neighbours, or
  - it has 2 neighbours and is currently alive.
  In any other case the cell will die."
  [n cell]
  (let [alive? (not (nil? cell))]
    (or (= n 3)
        (and (= n 2) alive?))))

(defn step
  "Produce the next generation."
  [cells]
  (let [m (frequencies (mapcat neighbours cells))
        lzseq (for [[cell n] m :when (will-live? n (cells cell))]
                cell)]
    (set lzseq)))

(def initial-cells #{[6 2] [6 3] [6 4] ;; blinker
                     [2 0] [3 1] [1 2] [2 2] [3 2] ; glider
                     [10 1] [9 2] [9 3] [11 2] [11 3] [10 4] ; beehive
                     })

(defn new-board
  []
  #{[2 0] [3 1] [1 2] [2 2] [3 2]})


src/game_of_life/dynamic.clj
(ns game-of-life.dynamic
  (:require
   [clojure.spec.alpha :as s]
   [game-of-life.conway :refer [step]]
   [quil.core :as q]))

(s/def ::generation pos-int?)

(s/def ::color number?)

(s/def ::cell (s/coll-of int?))
(s/def ::cells (s/coll-of ::cell))

(s/def ::num-cells-per-row pos-int?)
(s/def ::num-cells-per-column pos-int?)
(s/def ::grid (s/keys :req [::num-cells-per-row
                            ::num-cells-per-column]))

(s/def ::state (s/keys :req [::cells
                             ::color
                             ::generation
                             ::grid]))

(def zoom-factor 2)
(def min-rows 8)
(def min-columns 6)
(def max-rows 200)
(def max-columns 100)

;; TODO:
;; - random automata on mouse click
;; - double check num rows and num columns
;; - improve zoom
;; - use cljc and make a version for clojurescript
;; - save PDF when a button or a location on the screen is pressed
;; - use with-graphics macro
;; - use nicer font family: create-font? load-font?

(def initial-cells #{[6 2] [6 3] [6 4] ;; blinker
                     [2 0] [3 1] [1 2] [2 2] [3 2] ; glider
                     [10 1] [9 2] [9 3] [11 2] [11 3] [10 4] ; beehive
                     })

(defn setup-fn
  "setup function called only once, during sketch initialization."
  []
  (q/frame-rate 10)
  (q/color-mode :hsb)
  {::cells initial-cells
   ::color 0
   ::generation 1
   ::grid {::num-cells-per-row 10 ::num-cells-per-column 8}})

(defn update-fn
  "update function called on each iteration before the draw function."
  [state]
  (when-not (s/valid? ::state state)
    (throw (ex-info (s/explain-str ::state state)
                    (s/explain-data ::state state))))
  ;; (prn "STATE" state)
  (let [cells (step (::cells state))
        color (mod (+ (::color state) 0.7) 255)
        generation (inc (::generation state))]
    (-> state
        (assoc ::cells cells ::color color ::generation generation))))

(defn draw-grid
  [state]
  (when-not (s/valid? ::state state)
    (throw (ex-info (s/explain-str ::state state)
                    (s/explain-data ::state state))))
  (let [grid (::grid state)
        n-cw (::num-cells-per-column grid)
        n-ch (::num-cells-per-row grid)
        cw (quot (q/width) n-cw)
        ch (quot (q/height) n-ch)]
    (q/background 240)
    (q/fill 240)
    (q/stroke 0)
    (q/stroke-weight 1)

    (doseq [x (range n-cw)
            y (range n-ch)]
      (q/rect (* x cw) (* y ch) cw ch))))

(defn draw-automata
  [state]
  (let [grid (::grid state)
        n-cw (::num-cells-per-column grid)
        n-ch (::num-cells-per-row grid)
        cw (quot (q/width) n-cw)
        ch (quot (q/height) n-ch)]
    (q/fill (::color state) 255 255)
    (doseq [[x y] (::cells state)]
      (q/rect (* x cw) (* y ch) cw ch))))

(defn draw-label
  [state]
  (let [x (or (:x state) (/ (q/width) 2))
        y (or (:y state) 10)
        width 300
        height 100]
    (q/fill 255)
    (q/stroke-weight 1)
    (q/rect x y width height)
    (q/fill 0 0 0)
    (q/text-size 16)
    (q/text (format "Generation %d" (::generation state)) (+ 5 x) (+ 5 y) width height)
    (q/text (format "%d rows" (get-in state [::grid ::num-cells-per-row])) (+ 5 x) (+ 35 y) width height)
    (q/text (format "%d columns" (get-in state [::grid ::num-cells-per-column])) (+ 35 x) (+ 55 y) width height)))

(defn draw-fn
  "Draw function, i.e. the render loop."
  [state]
  (draw-grid state)
  (draw-automata state)
  (draw-label state))

(defn zoom-out [state]
  (let [nrows-old (get-in state [::grid ::num-cells-per-row])
        ncols-old (get-in state [::grid ::num-cells-per-column])
        nrows (min max-rows (int (Math/floor (* zoom-factor nrows-old))))
        ncols (min max-columns (int (Math/floor (* zoom-factor ncols-old))))]
    ;; (prn "zooming OUT" nrows-old nrows ncols-old ncols)
    (-> state
        (assoc-in [::grid ::num-cells-per-row] nrows)
        (assoc-in [::grid ::num-cells-per-column] ncols))))

(defn zoom-in [state]
  (let [nrows-old (get-in state [::grid ::num-cells-per-row])
        ncols-old (get-in state [::grid ::num-cells-per-column])
        nrows (max min-rows (int (Math/ceil (/ nrows-old zoom-factor))))
        ncols (max min-columns (int (Math/ceil (/ ncols-old zoom-factor))))]
    ;; (prn "zooming IN" nrows-old nrows ncols-old ncols)
    (-> state
        (assoc-in [::grid ::num-cells-per-row] nrows)
        (assoc-in [::grid ::num-cells-per-column] ncols))))

(defn mouse-moved [state event]
  (assoc state :x (:x event) :y (:y event)))

(defn mouse-wheel [state wheel-rotation]
  (if (= 1 wheel-rotation)
    (zoom-out state)
    (zoom-in state)))


src/game_of_life/ui.clj
(ns game-of-life.ui
  "Animation logic and rendering for the Game of Life."
  (:require
   [clojure.spec.alpha :as s]
   [game-of-life.conway :refer [initial-cells new-board step]]
   [quil.core :as q]))

(s/def ::color number?)
(s/def ::generation pos-int?)

(s/def ::cell (s/coll-of int?))
(s/def ::cells (s/coll-of ::cell))
(s/def ::cell-height pos?)
(s/def ::cell-width pos?)

(s/def ::aspect-ratio pos?)
(s/def ::columns pos-int?)
(s/def ::rows pos-int?)
(s/def ::zoom pos?)
(s/def ::grid (s/keys :req [::aspect-ratio
                            ::columns
                            ::rows
                            ::zoom]))

(s/def ::state (s/keys :req [::cells
                             ::cell-height
                             ::cell-width
                             ::color
                             ::generation
                             ::grid]
                       :opt [::mouse-x
                             ::mouse-y]))

(def zoom-min 0.2)
(def zoom-max 4.0)

(defn reset-board
  [state]
  (assoc state ::cells (new-board) ::color 0 ::generation 1))

(defn setup-fn
  "Setup (pure function). It's called only once, during sketch initialization.
  Return the initial state."
  []
  (q/frame-rate 10)
  (q/color-mode :hsb)
  (let [aspect-ratio (/ (q/width) (q/height))
        rows 10
        columns (q/floor (* rows aspect-ratio))
        cell-width (quot (q/width) columns)
        cell-height (quot (q/height) rows)]
    {::cells initial-cells
     ::cell-height cell-height
     ::cell-width cell-width
     ::color 0
     ::generation 1
     ::grid {::aspect-ratio aspect-ratio
             ::columns columns
             ::rows rows
             ::zoom 1}}))

(defn calc-rows
  "Euristic to set a reasonable number of rows.
  Other ideas:
  - given a minimum row height (e.g. 20px), compute the number of rows for a
   particular screen height (only once), i.e.:
   rows-min = (q/screen-height) / row-min-height
   then use this rows-min to compute the number of rows for the sketch (recompute
   it every draw loop), i.e.:
   rows-min : (q/screen-height) = x : (q/height)
   Finally, apply the zoom to x.
  - define multiple steps, e.g. [0 100) -> 10, [100 200) -> 20, etc."
  [zoom]
  (if (> (q/height) 400)
    (q/floor (* zoom 20))
    (q/floor (* zoom 10))))

(defn update-fn
  "Update function. Called at each iteration. Receives the old state and
  produces the new state. It runs before the draw function."
  [state]
  (when-not (s/valid? ::state state)
    (throw (ex-info (s/explain-str ::state state)
                    (s/explain-data ::state state))))
  (let [cells (step (::cells state))
        color (mod (+ (::color state) 0.7) 255)
        generation (inc (::generation state))
        ;; The sketch might have been resized, so we recompute the aspect ratio
        aspect-ratio (/ (q/width) (q/height))
        zoom (get-in state [::grid ::zoom])
        rows (calc-rows zoom)
        ;; Whenever the aspect ratio changes, we need to recompute the grid
        columns (q/floor (* rows aspect-ratio))
        grid (assoc (::grid state) ::aspect-ratio aspect-ratio ::columns columns ::rows rows)
        cell-width (quot (q/width) columns)
        cell-height (quot (q/height) rows)]
    (assoc state ::cells cells ::cell-width cell-width ::cell-height cell-height ::color color ::generation generation ::grid grid)))

(defn draw-grid
  [{cw ::cell-width ch ::cell-height {:keys [::columns ::rows]} ::grid}]
  (q/background 240)
  (q/fill 240)
  (q/stroke 0)
  (q/stroke-weight 1)
  (doseq [x (range columns)
          y (range rows)]
    (q/rect (* x cw) (* y ch) cw ch)))

(defn draw-automata
  [{cells ::cells cw ::cell-width ch ::cell-height color ::color}]
  (q/fill color 255 255)
  (doseq [[x y] cells]
    (q/rect (* x cw) (* y ch) cw ch)))

(defn draw-label
  [{generation ::generation grid ::grid mx ::mouse-x my ::mouse-y}]
  (let [x (or mx (/ (q/width) 2))
        y (or my 10)
        w 400
        h 100
        x-text (+ 5 x)]
    (q/fill 255)
    (q/stroke-weight 1)
    (q/rect x y w h)
    (q/fill 0 0 0)
    (q/text-size 16)
    (q/text (format "Generation %d" generation) x-text (+ 5 y) w h)
    (q/text (format "Grid %d x %d (aspect=%.2f zoom=%.2f)" (q/width) (q/height) (float (::aspect-ratio grid)) (float (::zoom grid))) x-text (+ 25 y) w h)
    (q/text (format "%d rows" (::rows grid)) x-text (+ 45 y) w h)
    (q/text (format "%d columns" (::columns grid)) x-text (+ 65 y) w h)))

(defn draw-fn
  "Draw the Game of Life (function with side-effects)."
  [state]
  (draw-grid (select-keys state [::cell-height ::cell-width ::grid]))
  (draw-automata (select-keys state [::cells ::cell-height ::cell-width ::color]))
  (draw-label (select-keys state [::mouse-x ::mouse-y ::generation ::grid])))

(defn zoom-out
  [state]
  (let [old-zoom (get-in state [::grid ::zoom])
        zoom (min zoom-max (* 2 old-zoom))]
    (assoc-in state [::grid ::zoom] zoom)))

(defn zoom-in
  [state]
  (let [old-zoom (get-in state [::grid ::zoom])
        zoom (max zoom-min (* 0.2 old-zoom))]
    (assoc-in state [::grid ::zoom] zoom)))

(defn mouse-clicked
  [state {button :button}]
  (if (= :left button)
    (reset-board state)
    state))

(defn mouse-moved
  [state {x :x y :y}]
  (assoc state ::mouse-x x ::mouse-y y))

(defn mouse-wheel
  [state wheel-rotation]
  (if (= 1 wheel-rotation)
    (zoom-out state)
    (zoom-in state)))

(defn key-pressed
  [state event]
  (if (= :i (:key event))
    (reset-board state)
    state))